### PR TITLE
Fix DisjunctiveConstraint edge case and add ConjunctiveDisjunctiveConstraint

### PR DESCRIPTION
**Pull request by boy2000-007man**
# What does this PR do?
- implement [`AC automaton`](https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm) to supersede Trie to fix DisjunctiveConstraint edge case
- add ConjunctiveDisjunctiveConstraint to handle the complex combinations between multiple conjunctive and disjunctive constraints
- update stronger unit tests

Fixes # (issue)

#17831 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten, @cwkeam

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hey @boy2000-007man,

Thanks for the fix proposal! @cwkeam could you take a look here as well? :-) 

@boy2000-007man - it'd be really nice if you could add a test that would have failed without your fix, but will now pass. 

Thanks a lot for working on this!

File: tests/generation/test_generation_beam_constraints.py [hunk 1]
```diff
@@ -53,63 +53,398 @@ def test_check_illegal_input(self):
...
+        self.assertTrue(dc.advance() == [3, 4, 5])
 
+        self.assertTrue(dc.does_advance(5))
         stepped, completed, reset = dc.update(5)
         self.assertTrue(dc.completed)  # Completed!
         self.assertTrue(dc.remaining() == 0)
         self.assertTrue(dc.current_seq == [1, 2, 5])
+        self.assertTrue(dc.advance() is None)
+
+    def test_dc_example_progression_mid_overlap_two(self):
```
Comment (boy2000-007man) [inline tests/generation/test_generation_beam_constraints.py:160]: @patrickvonplaten , here is the updated test case for the reported edge case

**Comment by patrickvonplaten** [comment]
Hey @boy2000-007man, 

Thanks a lot for the PR - I'm a bit worried about adding so much new code to main transformers to catch an edge case and wonder if it's really worth it. The problem is that this function will quickly become unmaintainable (it already sadly is to some extent) - in your opinion is it absolutely necessary to add this edge case? Also could you maybe provide a "real" generation example that shows how the current implementation fails?

**Comment by boy2000-007man** [comment]
Hi, @patrickvonplaten, the current code implementation is complex to support both the existing `DisjunctiveConstraint` and newly added `ConjunctiveDisjunctiveConstraint` at the same time. I can add a much-simplified version dedicated to back `DisjunctiveConstraint` only, and the new `ConjunctiveDisjunctiveConstraint` is not used by the library default but requires manual import, so won't break any existing works by chance.
Finding a failure case is not that straightforward especially without deep understanding of specific model preference, but I can image some constraints like `the small cat/small cats`, `the united states/united kingdom` may be influenced.

**Comment by patrickvonplaten** [comment]
Hey @boy2000-007man,

Sorry to reply so late here. Will gently ask @cwkeam in private if he could take a quick look because he's the most familiar with the current code. if there is no answer, I'll come back to it and dive deeper into the code to be able to better answer here. 

Also cc @gante if you're feeling curios on complex code ;-)

**Comment by gante** [comment]
Hi @boy2000-007man 👋 I was having a look into this PR, and one thing I noticed was that the objective of the PR was not immediately clear -- it says at the top that it fixes an edge case but... what edge case? We can find the answer to that in the code, especially in the docstring of `ConjunctiveDisjunctiveConstraint`.

I do think we should fix the edge case, as the documented behavior does not match the actual behavior. However, adding clear examples (as in #15761) will be extremely useful for our future selves 🙏 It will also helps the reviewers seeing the value of the PR :D

**Comment by boy2000-007man** [comment]
Hi, @gante , Sorry for the late reply. The edge case is mentioned in the associated bug report, https://github.com/huggingface/transformers/issues/17831. Do you mean to mention it again in the docstring?

**Pull request closed without merging**